File: app/src/main/java/com/esorokin/lantector/model/network/api/handler/SingleCallAdapter.java
package com.esorokin.lantector.model.network.api.handler;

import java.lang.reflect.Type;

import android.support.annotation.NonNull;

import io.reactivex.Scheduler;
import io.reactivex.Single;
import retrofit2.Call;
import retrofit2.CallAdapter;

class SingleCallAdapter<BaseResponse> implements CallAdapter<BaseResponse, Single<BaseResponse>> {
	private final Type responseType;
	private final NetworkErrorAdapter networkErrorAdapter;
	private final HttpResponseAdapter<BaseResponse> httpResponseAdapter;
	private final Scheduler scheduler;

	public SingleCallAdapter(@NonNull Type responseType,
	                         @NonNull NetworkErrorAdapter networkErrorAdapter,
	                         @NonNull HttpResponseAdapter<BaseResponse> httpResponseAdapter,
	                         @NonNull Scheduler scheduler) {
		this.responseType = responseType;
		this.networkErrorAdapter = networkErrorAdapter;
		this.httpResponseAdapter = httpResponseAdapter;
		this.scheduler = scheduler;
	}

	@Override
	public Type responseType() {
		return responseType;
	}

	@Override
	public Single<BaseResponse> adapt(Call<BaseResponse> call) {
		return new CallExecuteObservable<>(call)
				.singleOrError()
				.subscribeOn(scheduler)
				.onErrorResumeNext(throwable -> Single.error(networkErrorAdapter.adaptNetworkError(throwable)))
				.map(httpResponseAdapter::adaptHttpResponse);
	}
}

File: app/src/main/java/com/esorokin/lantector/model/network/api/handler/SingleCallAdapter.kt
package com.esorokin.lantector.model.network.api.handler

import io.reactivex.Scheduler
import io.reactivex.Single
import retrofit2.Call
import retrofit2.CallAdapter
import java.lang.reflect.Type

internal class SingleCallAdapter<BaseResponse>(
        private val networkScheduler: Scheduler,
        private val responseType: Type,
        private val networkErrorAdapter: NetworkErrorAdapter,
        private val httpResponseAdapter: HttpResponseAdapter<BaseResponse>) : CallAdapter<BaseResponse, Single<BaseResponse>> {

    override fun responseType(): Type {
        return responseType
    }

    override fun adapt(call: Call<BaseResponse>): Single<BaseResponse> {
        return CallExecuteObservable(call).singleOrError()
                .subscribeOn(networkScheduler)
                .onErrorResumeNext { throwable -> Single.error(networkErrorAdapter.adaptNetworkError(throwable)) }
                .map(httpResponseAdapter::adaptHttpResponse)
    }
}
